subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -110,36 +136,106 @@ private boolean experimentalCompilationAvoidanceEnabled() {
     }
 
     @Override
-    @TaskAction
     protected void compile() {
+        compile(null);
+    }
+
+    @TaskAction
+    protected void compile(InputChanges inputChanges) {
         checkGroovyClasspathIsNonEmpty();
         warnIfCompileAvoidanceEnabled();
+
+        this.sourceClassesMappingFile = new File(getTemporaryDir(), "source-classes-mapping.txt");
+
+        Multimap<File, String> oldMappings = readSourceClassesMappingFile(sourceClassesMappingFile);
+        doCompile(inputChanges, oldMappings);
+        updateSourceClassesMappingFile(inputChanges, oldMappings);
```

Comment:
Let's not always update the mappings file. Only do it when incremental compilation is enabled (options.incremental).